languages/ruby/reference/exercise-concepts/nucleotide-count.md

```diff
@@ -0,0 +1,44 @@
+# Concepts of nucleotide-count
+
+## General
+
+- class methods: used as the starting point of the exercise. They can be named with `self.[method_name]` notation or with a `self` block.
+- module methods: depending on the approach, they can be used instead of the class methods.
+- factory pattern and object creation: the method `from_dna` can be used to create instances of the main class.
+- instance variables: @variables are needed to store the common value/state for all object methods.
+- accessors: `att_reader` can be used to access instance variables.
+- access modifiers: `private` can be used to restrict access to the readers.
+- method definition: the `histogram` method needs to be defined.
+- method arguments: the method `from_dna` needs an argument.
+- string: one string is passed as the input.
+- chars: it's necessary to convert the string into an array of chars.
+- strings mutation: `freeze` can be used to prevent string constants from mutating.
+- Enumerable module: including `Enumerable` and implement `each` is one of the possible approaches to this exercise.
+- Enumerable#count: one of the possible ways of counting nucleotides.
+- Enumerable#inject vs Enumerable#each_with_object.
+- exceptions: necessary to raise an exception if the input is not valid.
+- return values: methods need to return values either implicitly or explicitly.
+- guard clauses: a guard clause can be used to raise an exception.
+- constants: constants can be used to store static information like DNA values.
+- namespaces: namespaces can be used along with constants.
+- if conditions: an if condition can be used to check if the input is valid.
+- yield: in case `Enumerable` is included the `each` method has to use `yield`.
+- blocks: they are used in any of the Enumerable methods.
+- Ruby magic comments: `frozen_string_literal` can be used.
+- require: use to import test library and `nucleotide_count` file
```

Comment:
Is this really a learning goal for the student? I don't think that requiring files is part of this exercise. The test suite has some require expressions, but the `nucleotide_count.rb` file written by the student doesn't need to require anything.

Reply:
Now I'm also tempted to add Minitest to the concepts 😄 I think the student also needs to understand the code on the test files. What are your thoughts on this?

Reply:
The test files are something that must be idiomatic, exemplary code because it is delivered to the student.  But the student should not be concerned with the things there, unless the exercise is specifically geared toward teaching testing.

Since the test code is POR<sup>1</sup> the concepts used in the tests will hopefully be covered in the concept exercises.

---
1: Plain Old Ruby

Reply:
Hmmm, I forgot that these exercises are not only solved in the browser :/ Maybe we *should* have a `minitest` concept as the students are required to comment out the `skip`s, so they'll need at least rudimentary understanding of if.

Reply:
But then again, this would be a concept that applies to almost every exercise, so does it really make sense to explicitly have it as a pre-requisite?

Reply:
Yeah, I think for the web based concept files, they won't see the tests but for the practice ones they will.  And for that, I hope that the "Hello-World" exercise gives those instructions.

There have been discussions about having an exercise where the focus _is_ on testing, though.  For now, I don't think we are worrying about that.

Reply:
Thanks for your comments. It's now clear to me that it's not part of the concepts for this exercise. As you @goalaleo mention, maybe it's nice to say something about testing in a more general level but not here.